### ISSUE TITLE
Not working on NUMA system

### ISSUE THREAD

**Issue by marki555** [fireice-uk/xmr-stak-cpu#251]
Hi,
I have tried the latest release with hwloc/numa support on 8 socket system, but it doesn't work. Older version works without issues.
It seems like only one thread out of 176 are hashing (as per `h` report), `top` shows only 200% CPU usage for the process, but load average is about 170.
What is strange is 47 seconds delay between launching last thread and first memory pinning.

```
[2017-07-26 11:13:29] : Starting single thread, affinity: 272.
[2017-07-26 11:13:29] : Starting single thread, affinity: 273.
[2017-07-26 11:13:30] : Connecting to pool localhost:23456 ...
[2017-07-26 11:13:30] : Connected. Logging in...
[2017-07-26 11:14:17] : hwloc: memory pinned
[2017-07-26 11:14:18] : hwloc: memory pinned
```

**Comment by Dead2** [comment]
If I remember correctly, xmr-stak-cpu can only handle 64threads. What you could try is to run one instance of xmr-stak-cpu per socket, or per two sockets.
Big NUMA systems like that can exhibit peculiarities with memory allocation, and you might be triggering one of those. Do you have memory connected to each socket?

I am assuming linux here, if so, you might try to run the following that might or might not help:
sysctl -w vm.nr_overcommit_hugepages=4000
That lets the kernel attempt to find new hugepages other places than the ones that are preallocated by vm.nr_hugepages. That might possibly help if the problem is that it has trouble allocating memory from the correct socket, and thus have to use slow memory elsewhere.
vm.nr_hugepages should also be set to at least as much as the number of threads you want to run, but not too much more, since that ram cannot be used for anything else. And with overcommit enabled, it allows a lot more flexibility.

Is this an Intel system? If so, remember that each thread uses 2MB cache, so find out how much L3 cache each socket has, and run as many as that allows. Exactly what cores within that socket has access to the cache can be tricky, since L3 cache might be partitioned on the bigger cpus. Try to optimize for one socket first, and once you find a good setup, just duplicate that for each of the other sockets as well.

When you have tried this, and If you want more help, you'll need to post cpu model, OS, memory and config file. Also, the output from  "numactl -m" both before and while xmr-stak-cpu is running can help with debugging numa-related problems. "lshw -short" is also useful.

**Comment by psychocrypt** [comment]
xmr-stak should be able to handle more than 64 threads.
Please post the cpu type and the config which is suggested by the miner.

**Comment by marki555** [comment]
Older version (pulled from git few weeks ago) works OK, but threads suggestion there was not very good. This is Intel E7-8880, which has 45 MB L3 cache and 18 cores (and RAM is installed to all sockets). I am running the older version with 22 threads per socket, this has given me the highest hashrate.
I assume the new NUMA version is not using slow memory as otherwise it would complain about that.

What options exactly to pass to `numactl`?

Running `hwloc-ps -t` shows all threads bind to PU:2, only one to PU:212 (probably the one which actually hashes). When running the older non-NUMA version, it shows correctly each thread bind to different PU and the process itself, 1st thread and last 2 threads bind to "Machine:0".

`lstopo` looks like this:
```
Machine (512GB total)
  NUMANode L#0 (P#0 64GB)
    Socket L#0 + L3 L#0 (45MB)
      L2 L#0 (256KB) + L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0
        PU L#0 (P#0)
        PU L#1 (P#144)
      L2 L#1 (256KB) + L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1
        PU L#2 (P#1)
        PU L#3 (P#145)
      L2 L#2 (256KB) + L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2
        PU L#4 (P#2)
        PU L#5 (P#146)
```

**Comment by psychocrypt** [comment]
please post the suggested config from the current release.

**Comment by Dead2** [comment]
@marki555 I meant "numastat -m", sorry.

You should post your thread-config and what OS/distro you are running.
The fact that it binds to only one core is strange though, but it would explain the performance problems, both the slow hashrate and the slow startup. Not sure why it does not respect your affinity settings though. It could be an os-level override, but it seems a bit unlikely.

@psychocrypt I thought @fireice-uk said there was some problem with nonce splitting on more than 64 threads, but I might be remembering wrong.

**Comment by psychocrypt** [comment]
@Dead2 I said this on reddit and fixed the issue, but the tests on a 80 core system we refactored the detection and can't test again on such a system. Therefore I need to see the suggeated config.

**Comment by Dead2** [comment]
@psychocrypt Great 👍

**Comment by psychocrypt** [comment]
@marki555 Please post the full output of lstopo and the suggested config

**Comment by marki555** [comment]
Attached both suggested config (it suggests one more thread per socket compared with my config) and lstopo. Distro is SLES11SP3 (a bit old, but that is what friend's uni uses for their science computations).

BTW should there be any measurable difference in hash speed with the new version (with low_power_mode false)? Otherwise we are fine with running the older version, just wanted to test it out. Also if you have anything to test on NUMA system, we can (when there are not any other tasks running there).

[lstopo.txt](https://github.com/fireice-uk/xmr-stak-cpu/files/1177784/lstopo.txt)
[sugg_config.txt](https://github.com/fireice-uk/xmr-stak-cpu/files/1177785/sugg_config.txt)

**Comment by marki555** [comment]
I have made tests with few threads only... Here are the results of `hwloc-ps -t`:
```
# 1x 2 threads (means 2 threads on socket0)
39253   Machine:0               xmr-stak-cpu
 39253  Machine:0
 39254  Machine:0
 39255  PU:0
 39256  Machine:0
 39257  Machine:0
 39258  Machine:0

# 2x 2 threads
39323   Machine:0               xmr-stak-cpu
 39323  Machine:0
 39324  Machine:0
 39326  Machine:0
 39327  PU:2
 39328  Machine:0
 39329  Machine:0
 39330  Machine:0
 39331  Machine:0

# 2x 1 threads
39449   Machine:0               xmr-stak-cpu
 39449  Machine:0
 39450  Machine:0
 39451  Machine:0
 39452  PU:36
 39453  Machine:0
 39454  Machine:0

# 8x 1 threads (each socket 1 thread)
40101   PU:149 PU:151 PU:252            xmr-stak-cpu
 40101  PU:151
 40103  PU:151
 40104  PU:151
 40105  PU:151
 40106  PU:151
 40107  PU:252
 40108  PU:252
 40109  PU:252
 40110  PU:149
 40111  PU:149
 40112  PU:149
 40113  PU:149
# few seconds later...
40101   Machine:0               xmr-stak-cpu
 40101  Machine:0
 40103  Machine:0
 40104  Machine:0
 40105  Machine:0
 40106  PU:72
 40107  Machine:0
 40108  Machine:0
 40109  Machine:0
 40110  Machine:0
 40111  Machine:0
 40112  Machine:0
 40113  Machine:0

# 1x 2 threads + 7x 1 threads (first socket 2 threads, all other sockets 1 thread)
40181   PU:1 PU:36              xmr-stak-cpu
 40181  PU:1
 40182  PU:1
 40183  PU:1
 40184  PU:1
 40185  PU:36
 40186  PU:1
 40187  PU:1
 40188  PU:1
 40189  PU:1
 40190  PU:1
 40191  PU:1
 40192  PU:1
 40193  PU:1
```

Strange is that with few threads, they are not bound to specific PU, only to Machine:0 and hashing works on all of them. `hwloc-ps` shows no output when threads are started, only after the memory pinned message. Interesting output was captured when running 1 thread on each cpu socket, as the binding changed from PU to Machine.

**Comment by psychocrypt** [comment]
Could be that the bug is related to #254. It was a racecondition in the current release. Could you please test the current  dev.

> BTW should there be any measurable difference in hash speed with the new version (with low_power_mode false)? 

Yes we measured a small speed increase.

**Comment by marki555** [comment]
This is the result with dev branch. Still not working, but now 10 threads managed to set affinity correctly instead of just one thread.

```
# ./hwloc-ps -t|awk '{print $2}'|sort|uniq -c
      1 PU:148
      1 PU:150
      1 PU:178
      1 PU:192
      1 PU:194
      1 PU:202
      1 PU:216
      1 PU:224
      1 PU:232
      1 PU:256
    171 PU:47
```

**Comment by psychocrypt** [comment]
If you compile with `-DHWLOC_ENABLE=OFF` and disable HWloc than this issue should be solved. see issue #256.

**Comment by marki555** [comment]
Yes, compiling without hwloc works OK as I guess it behaves like the previous version (which also works ok). But shouldn't hwloc provide some advantage on big NUMA systems?

**Comment by fireice-uk** [comment]
@marki555 That's correct. However it looks like hwloc is not doing what we want it to do on some systems.

@psychocrypt Is there any way we can get hold of a test system where hwloc is not behaving correctly? My dual socket systems with Arch are a-ok.

**Comment by psychocrypt** [comment]
@fireice-uk I can't reproduce the bug on my dual cor laptop. I will check if I can reproduce  it on another system.

**Comment by marki555** [comment]
@psychocrypt, @fireice-uk just let me know what needs to be tested or what info provided and I can run it on our system...

**Comment by fireice-uk** [comment]
@marki555 any chance you can set me up with a user shell account and ssh? It will be much faster than walking you through gdb usage.

**Comment by marki555** [comment]
@fireice-uk sorry, no access possible from outside due to firewalls. But I have some basic gdb knowledge. Or you could make a version which would print some debugging info to console.

**Comment by fireice-uk** [comment]
@marki555 Ok. Can you post the result of `uname -r`? I will make a PR either at the end of this week or early next week that might fix it.

**Comment by marki555** [comment]
3.0.101-0.47.71-default
